### PR TITLE
Scrollable full screen dialog

### PR DESCRIPTION
**Pull request by andrewjtait**
* Makes `DialogFullScreen` scrollable if the content goes beyond the height of the browser.
* Adds `Pages` and `Page` components to enable paginated content for full screen views.

```
const headingOne = (
  <Heading title="My First Page" />
);

const headingTwo = (
  <Heading title="My Second Page" />
);

return (
  <DialogFullScreen>
    <Pages slideIndex={ 0 }>
      <Page title={ headingOne }>
        Content for the first page.
      </Page>

      <Page title={ headingTwo }>
        Content for the second page.
      </Page>
    </Pages>
  </DialogFullScreen>
);
```

### CODE REVIEW

File: src/components/pages/page/page.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import PropTypes from 'prop-types';
+import tagComponent from './../../../utils/helpers/tags';
+import FullScreenHeading from './../../dialog-full-screen/full-screen-heading';
+
+const pageClasses = (props) => {
+  return classNames('carbon-page', props.className);
+};
+
+const Page = props =>
+  <div className={ pageClasses(props) } { ...tagComponent('page', props) }>
```
Comment (ianoxley) [inline src/components/pages/page/page.js]: An `<article>` tag here would be more semantic than a `<div>`.

File: src/utils/helpers/browser/browser.js [hunk 2]
```diff
@@ -162,6 +172,43 @@ const Browser = {
...
+
+  /**
+   * Determines if a class is present on an element.
+   *
+   * @param {Element} elem - the element to inspect
+   * @param {String} name - the class name to look for
+   * @return {Boolean}
+   */
+  hasClass: (elem, name) => {
+    return new RegExp(classRegex(name)).test(elem.className);
```
Comment (ianoxley) [inline src/utils/helpers/browser/browser.js]: You could use `elem.classList.contains(name)` here instead.

File: src/utils/helpers/browser/browser.js [hunk 3]
```diff
@@ -162,6 +172,43 @@ const Browser = {
...
+   * Removes a class from an element if it exists.
+   *
+   * @param {Element} elem - the element to remove the class from
+   * @param {String} name - the class name to remove
+   * @return {Boolean}
+   */
+  removeClass: (elem, name) => {
+    if (Browser.hasClass(elem, name)) {
+      elem.className = elem.className.replace(new RegExp(classRegex(name), 'g'), ' ');
+    }
```
Comment (ianoxley) [inline src/utils/helpers/browser/browser.js]: You could use `elem.classList.remove(name)` instead.

File: src/utils/helpers/browser/browser.js [hunk 4]
```diff
@@ -162,6 +172,43 @@ const Browser = {
...
+   * Adds a class to an element if it is not already present.
+   *
+   * @param {Element} elem - the element to add the class to
+   * @param {String} name - the class name to add
+   * @return {Boolean}
+   */
+  addClass: (elem, name) => {
+    if (!Browser.hasClass(elem, name)) {
+      elem.className = elem.className ? [elem.className, name].join(' ') : name;
+    }
```
Comment (ianoxley) [inline src/utils/helpers/browser/browser.js]: You could use `elem.classList.add(name)` instead.

File: src/components/dialog-full-screen/dialog-full-screen.js [hunk 5]
```diff
@@ -91,16 +103,44 @@ class DialogFullScreen extends Modal {
   }
 
   /**
-   * Returns title prop wrapped in <h2> if title is a string otherwise returns the title prop directly.
-   * Dialog main classes.
+   * Overrides the original function to disable the document's scroll.
+   */
+  get onOpening() {
+    Browser.addClass(this.document.body, DIALOG_OPEN_BODY_CLASS);
```
Comment (ianoxley) [inline src/components/dialog-full-screen/dialog-full-screen.js]: Do we need a `Browser.addClass` function? We could just use `this.document.body.classList.add(DIALOG_OPEN_BODY_CLASS)`.

And is setting a class the right way to go here? e.g. as the `<body>` isn't actually the dialog element. Would using a data attribute be better? e.g. `data-fullscreendialogopen`.

File: src/components/dialog-full-screen/dialog-full-screen.js [hunk 6]
```diff
@@ -91,16 +103,44 @@ class DialogFullScreen extends Modal {
...
+  }
+
+  /**
+   * Overrides the original function to enable the document's scroll.
+   */
+  get onClosing() {
+    Browser.removeClass(this.document.body, DIALOG_OPEN_BODY_CLASS);
```
Comment (ianoxley) [inline src/components/dialog-full-screen/dialog-full-screen.js]: We could use `this.document.body.classList.remove(DIALOG_OPEN_BODY_CLASS)` instead.

File: src/components/dialog-full-screen/dialog-full-screen.js (hunk 5 above)
Comment (andrewjtait) [inline src/components/dialog-full-screen/dialog-full-screen.js]: the data attribute wouldn't trigger the CSS redraw properly in IE which is why I updated it to use a class instead (I guess using a class IE is more liberal with CSS updates).

Keeping the helper function allows us to check if the class already exists more easily so we don't get duplicate classes.

File: src/utils/helpers/browser/browser.js (hunk 4 above)
Comment (andrewjtait) [inline src/utils/helpers/browser/browser.js]: as long as it is properly supported across all supported browsers 👍

File: src/utils/helpers/browser/browser.js (hunk 2 above)
Comment (andrewjtait) [inline src/utils/helpers/browser/browser.js]: as long as it is properly supported across all supported browsers 👍

File: src/components/dialog-full-screen/dialog-full-screen.js (hunk 5 above)
Comment (ianoxley) [inline src/components/dialog-full-screen/dialog-full-screen.js]: When you use `elem.classList.add(name)` you don't need to check whether the class has already been added: if it has, it doesn't get added again.

File: src/utils/helpers/browser/browser.js (hunk 2 above)
Comment (ianoxley) [inline src/utils/helpers/browser/browser.js]: It's pretty well supported: https://caniuse.com/#feat=classlist

File: src/components/dialog-full-screen/dialog-full-screen.js (hunk 5 above)
Comment (alandomiel) [inline src/components/dialog-full-screen/dialog-full-screen.js]: 👍  I checked here for both `add` and `remove` https://www.w3schools.com/jsref/prop_element_classlist.asp

**Comment by shauncox1** [comment]
Passed Q/A see https://jira-sage.valiantyscloud.net/secure/ViewSession.jspa?testSessionId=42657 for details

**Comment by alandomiel** [comment]
test session : https://jira-sage.valiantyscloud.net/secure/ViewSession.jspa?testSessionId=42657

**Comment by ianoxley** [comment]
This resolves #1343

**Pull request merged**